src/gasman.c

```diff
@@ -484,6 +450,39 @@ static inline UInt IS_BAG_BODY(void * ptr)
             ((UInt)ptr & (sizeof(Bag) - 1)) == 0);
 }
 
+#if defined(MEMORY_CANARY)
+
+#include <valgrind/valgrind.h>
+#include <valgrind/memcheck.h>
+Int canary_size() {
+  Int bufsize = (Int)EndBags - (Int)AllocBags;
+  return bufsize<4096?bufsize:4096;
+}
+
+void ADD_CANARY() {
+  VALGRIND_MAKE_MEM_NOACCESS(AllocBags, canary_size());
+}
+void CLEAR_CANARY() {
+  VALGRIND_MAKE_MEM_DEFINED(AllocBags, canary_size());
+}
+#define CANARY_DISABLE_VALGRIND()  VALGRIND_DISABLE_ERROR_REPORTING
+#define CANARY_ENABLE_VALGRIND() VALGRIND_ENABLE_ERROR_REPORTING
+
+void CHANGED_BAG(Bag bag) {
+    CANARY_DISABLE_VALGRIND();
```

Comment:
Optional: if you move this around, you could also run `git clang-format` on your commit